param/parameterized.py

```diff
@@ -58,6 +60,51 @@
 VERBOSE = INFO - 1
 logging.addLevelName(VERBOSE, "VERBOSE")
 
+
+def _deprecate_positional_args(func):
+    """Internal decorator for methods that issues warnings for positional arguments
+
+    Using the keyword-only argument syntax in pep 3102, arguments after the
+    ``*`` will issue a warning when passed as a positional argument.
+
+    Adapted from scikit-learn
+    """
+    signature = inspect.signature(func)
+
+    pos_or_kw_args = []
+    kwonly_args = []
+    for name, param in signature.parameters.items():
+        if param.kind in (inspect.Parameter.POSITIONAL_OR_KEYWORD, inspect.Parameter.POSITIONAL_ONLY):
+            pos_or_kw_args.append(name)
+        elif param.kind == inspect.Parameter.KEYWORD_ONLY:
+            kwonly_args.append(name)
+            if param.default is inspect.Parameter.empty:
+                raise TypeError("Keyword-only param without default disallowed.")
+
+    @functools.wraps(func)
+    def inner(*args, **kwargs):
+        name = func.__qualname__.split('.')[0]
+        n_extra_args = len(args) - len(pos_or_kw_args)
+        if n_extra_args > 0:
+            extra_args = ", ".join(kwonly_args[:n_extra_args])
+
+            warnings.warn(
+                f"Passing '{extra_args}' as positional argument(s) to 'param.{name}' "
+                "was deprecated, please pass them as keyword arguments.",
```

Comment:
```suggestion
                "is deprecated and will stop working in Param X.X. Please pass them as keyword arguments.",
```

Reply:
Yes not mentioning the exact Param release is intentional. I think Param 2.0 took a long time in the making because some changes were announced to happen specifically in Param 2.0 and it took a long time to happen. Oldest mention I can find on Github is in https://github.com/holoviz/param/issues/73#issuecomment-90621960 and it is 8 years old! cc @jlstevens and it's from him and it's kind of funny :)

In practice what I envision is that this would emit a `DeprecationWarning` and about 3 to 6 months before the change (in Param 3.0 I guess) the warning emitted would be a `FutureWarning` so that all users get it.

Reply:
Most people would not know that 2.0 took a long time and will just install the package. 

And my feeling is that most people are lazy, and without a real deadline, there is no urgency and will therefore postpone it.

Reply:
> Most people would not know that 2.0 took a long time and will just install the package.

I agree.

Simon's suggestion also seems sensible: giving the relevant param version in the warning.

> Oldest mention I can find on Github is in #73 and it is 8 years old!

But is 8 years enough? :-)

Reply:
Here are a couple counter-arguments.

I have been bitten enough by having to support Python 2 (syntax, test suite, etc.) for such a long time, because it was somehow announced, that I would now be very careful with announcing anything concrete. To me, announcing something is deprecated and programmatically is the most important piece of information. Having the version in the warning message also sets a precedent, making it more likely to have to add it to other warnings, reinforcing the probability for us to tie our hands more than we would like to, while I believe not providing much value to Param users.

Yes, people are lazy, and they usually know it's a bad thing! I would **never** complain to a maintainer for a breaking change, for which a warning was emitted for a long period of time, I'll just blame myself for being too lazy. All the more so when that breaking change happens in a major release, which we would do here.

I would not be so picky if there was not history, but I bet that if asked 8 years ago, none of the Param maintainers would have guessed that it would have taken so long for the release to happen (it's not over by the way, who knows how long it is going to take for it to really happen!).